lib/consumer.js

```diff
@@ -5,82 +7,77 @@ var amqp = require('amqplib'),
+          amqpConnection = null;
+          amqpChannel = null;
+        });
 
-        amqpConnection.on('close', reconnect);
-        amqpConnection.on('error', reconnect);
+        _connection.on('error', Logger.error);
 
-        return amqpConnection.createChannel();
+        amqpConnection = _connection;
+        resolve(amqpConnection);
       })
-      .then(function (_channel) {
-        amqpChannel = _channel;
-        amqpChannel.prefetch(amqpConfig.amqpPrefetch);
+      .catch((err) => {
+        amqpConnection = null;
+        reject(err);
+      });
+    });
+  }
+
+  return amqpConnection;
+};
 
-        Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
+var createChannel = () => {
+  if (!amqpChannel) {
+    amqpChannel = new Promise((resolve, reject) => {
+      amqpConnection.createChannel()
+        .then((_channel) => {
+          _channel.prefetch(amqpConfig.amqpPrefetch);
 
-        amqpState = amqpStates.CONNECTED;
+          Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
 
-        amqpIntervalId = clearInterval(amqpIntervalId);
+          _channel.on('close', (err) => {
+            if (err) Logger.error('ERROR:', err);
+            amqpChannel = null;
+          });
 
-        amqpChannel.on('close', reconnect);
-        amqpChannel.on('error', reconnect);
+          _channel.on('error', Logger.error);
 
-        while (amqpConnPool.length > 0) {
-          amqpConnPool.shift()(amqpChannel);
-        }
+          amqpChannel = _channel;
+          resolve(amqpChannel);
 
-        if (amqpReconnection) {
-          amqpReconnection = false;
-          for (var i = 0, l = amqpConsumptions.length; i < l; ++i) {
-            consume(amqpConsumptions[i].queue, amqpConsumptions[i].options, amqpConsumptions[i].callback);
+          if (amqpReconnect) {
+            amqpReconnect = false;
+            for (var i = 0, l = amqpQueues.length; i < l; ++i) {
+              consume(amqpQueues[i].queue, amqpQueues[i].options, amqpQueues[i].callback);
+            }
           }
-        }
+        })
+        .catch((err) => {
+          amqpChannel = null;
+          reject(err);
+        });
+    });
+  }
 
-        _resolve(amqpChannel);
-      })
-      .catch(reconnect);
-    } else if (amqpState === amqpStates.CONNECTING) {
-      amqpConnPool.push(_resolve);
-    } else {
-      _resolve(amqpChannel);
-    }
-  });
-};
-var reconnect = function(err) {
-  if (err) Logger.error('[BMQ-CONSUMER]', err);
-
-  amqpIntervalId = (!amqpIntervalId) ? clearInterval(amqpIntervalId) : amqpIntervalId;
-  Logger.info('[BMQ-CONSUMER] Try to reconnect');
-  amqpState = amqpStates.DISCONNECTED;
-  amqpReconnection = true;
-  amqpIntervalId = setInterval(connect, 1000);
+  return amqpChannel;
 };
 
-var checkRpc = function(msg, queue) {
+var checkRpc = (msg, queue) => {
   return function (_content) {
     if (_content !== undefined && msg.properties.replyTo) {
```

Comment:
I wouldn't say that you are wrong if you can explain it and it makes sense :-)  I guess it is the matter of applied way of doing things (according). It would equally made sense, in you example, for it to return an error (e.g. UserNotFoundError) which you can then decide and implement how you wish to handle it in the .catch handler if that was the accord. But I guess it boils down to the same, you either can check for null in then or handle in the catch. Just a matter of accord in the project. :-)